Expose adjustWhitespace to TextEditor API.

rgrunber opened the pull request:
- Disable adjustWhitespace in text edits with snippets so that code actions do not adjust snippet indentation by default
- Add testcase
- Fixes #63129 

Based on comments like https://github.com/microsoft/language-server-protocol/pull/1892#issuecomment-1964186341 , snippet edits should probably not be adjusting whitespace by default. For language support, once workspace edits have a property like, https://github.com/microsoft/vscode/blob/275faf6f08b7aa50843f3c18406b4d5969784e52/src/vscode-dts/vscode.d.ts#L4996-L5001 we could set the snippet insertion to respect that.

**Before**

https://github.com/user-attachments/assets/7e1c4819-e127-45f5-82e9-5bede1c26d03

**After**

https://github.com/user-attachments/assets/ae65bbcf-af10-4103-9fc4-31165aefa59f

rgrunber committed:
Expose adjustWhitespace to TextEditor API.

- Disable adjustWhitespace in text edits with snippets so that code
  actions do not adjust snippet indentation by default
- Adjust Emmet extension so that adjustWhitespace is enabled
- Add testcase

Signed-off-by: Roland Grunberg <rgrunber@redhat.com>

rgrunber commented:
@microsoft-github-policy-service agree [company="Red Hat"]

jrieken commented on src/vscode-dts/vscode.d.ts:
```diff
@@ -1312,6 +1312,10 @@ declare module 'vscode' {
+			/**
+			 * Adjust the indentation of the snippet.
+			 */
+			readonly adjustWhitespace?: boolean;
```
In the API optional properties default to `false` and that's not the case here and also undesirable (we cannot change the existing behaviour here). We would need a name that's the other way around, I suggest `keepWhitespace` which is already used for `CompletionItem` for the same/similar purpose

jrieken replied:
to double-click. Only change the API name and do the conversion in the API implementation [here](https://github.com/microsoft/vscode/blob/0c7bdaaae9b654d97ee5b212bbf4b81bd871540f/src/vs/workbench/api/common/extHostTextEditor.ts#L524)

rgrunber replied:
I have addressed this (I think). I didn't know `extHostTextEditor` ends up calling `mainThreadEditor` but that makes more sense now. So `keepWhitespace` is optional, and when undefined, we set it to `false` (current default behaviour).

jrieken commented on src/vs/workbench/contrib/bulkEdit/browser/bulkTextEdits.ts:
```diff
@@ -156,7 +156,7 @@ class EditorEditTask extends ModelEditTask {
 					}
 				}
-				snippetCtrl.apply(snippetEdits, { undoStopBefore: false, undoStopAfter: false });
+				snippetCtrl.apply(snippetEdits, { undoStopBefore: false, undoStopAfter: false, adjustWhitespace: false });
```
This seems like an unrelated and unwanted change

rgrunber replied:
Just to explain why I added this. Yes, the main change is about adding an optional flag to `insertSnippet`, but I noticed that when a provided code action is applied, it doesn't go through the text editor `insertSnippet`. It seems to directly call the `SnippetController` here. This is what affects the actual behaviour for all extensions providing code actions. So any extension with a code action provider that includes snippets (eg. https://github.com/redhat-developer/vscode-java/issues/3686), and don't want VS Code re-adjusting indentation won't be resolved without this change. I also thought I could change the default behaviour (adjust indentation for snippets) based on comments like https://github.com/microsoft/language-server-protocol/pull/1892#issuecomment-1964186341 .

If this change to the default isn't wanted, would you support (in this PR) adding a field to `vscode.CodeAction`, `keepWhitespace`, much like is done for `vscode.CompletionItem`, and have it be honoured at https://github.com/microsoft/vscode/blob/12b060f0a812fcd7e4cabf9c65e8267f48519459/src/vs/editor/contrib/codeAction/browser/codeAction.ts#L299-L306 .

jrieken replied:
I understand but also here we cannot change the existing behaviour. I think this would be `keepWhitespace` for the `SnippetTextEdit` type

rgrunber replied:
Sorry for the delay. I've removed this line from the `snippetCtrl.apply(..)` call in `bulkTextEdits`. I've introduced an optional property `keepWhitespace`, for `SnippetTextEdit`. If set, it and it will end up passed down into the snippet session logic, where it will influence `adjustIndentation`. If not set, then the default behaviour wins out (indentation is adjusted).

So to summarize we have 2 ways to influence the indentation of a code action with a snippet text edit. As a separate option passed into `TextEditor.insertSnippet(..)` or as an option embedded in the `SnippetTextEdit` so that it works through the language provider.

jrieken replied:
Nice. Thanks for that!

jrieken commented on extensions/emmet/src/abbreviationActions.ts:
```diff
@@ -636,7 +636,7 @@ async function expandAbbreviationInRange(editor: vscode.TextEditor, expandAbbrLi
 			const expandedText = expandAbbr(expandAbbrInput);
 			if (expandedText) {
-				await editor.insertSnippet(new vscode.SnippetString(expandedText), expandAbbrInput.rangeToReplace, { undoStopBefore: false, undoStopAfter: false });
+				await editor.insertSnippet(new vscode.SnippetString(expandedText), expandAbbrInput.rangeToReplace, { undoStopBefore: false, undoStopAfter: false, adjustWhitespace: true });
```
This isn't needed when `keepWhitespace` is used as API

jrieken commented on extensions/emmet/src/abbreviationActions.ts:
```diff
@@ -650,7 +650,7 @@ async function expandAbbreviationInRange(editor: vscode.TextEditor, expandAbbrLi
 	const allRanges = expandAbbrList.map(value => value.rangeToReplace);
 	if (expandedText) {
-		return editor.insertSnippet(new vscode.SnippetString(expandedText), allRanges);
+		return editor.insertSnippet(new vscode.SnippetString(expandedText), allRanges, { undoStopBefore: true, undoStopAfter: true, adjustWhitespace: true });
```
Same, not needed when having `keepWhitespace`

jrieken reviewed (changes requested):
Thanks for this

rgrunber committed:
Respond to review comments.

- Rename adjustWhitespace to keepWhitespace
- Retain default behaviour of adjusting whitespace when keepWhitespace
  not defined

Signed-off-by: Roland Grunberg <rgrunber@redhat.com>

rgrunber committed:
Remove incorrect method parameter default.

Signed-off-by: Roland Grunberg <rgrunber@redhat.com>

rgrunber committed:
Introduce 'keepWhitespace' property for snippet text edit abstractions.

- Restore original behaviour of bulk text edits adjusting whitespace
- Add testcase

Signed-off-by: Roland Grunberg <rgrunber@redhat.com>

jrieken commented:
Also asking @mjbvz to review because this adds new API without going the proposed api flow

superle3 commented:
@jrieken  https://github.com/microsoft/vscode/issues/214757 is not a duplicate of this issue.
This issue is about whitespace, while https://github.com/microsoft/vscode/issues/214757 is about snippets not merging. I tested the issue on the latest commit (25418c6a590442002093b95e8764e59c551e7712) and it still persists. Could you please reopen that issue.
